contracts/clients/src/BlsSigner.ts

```diff
@@ -0,0 +1,215 @@
+    };
+
+    const nonce = await BlsWalletWrapper.Nonce(
+      this.wallet.PublicKey(),
+      this.verificationGatewayAddress,
+      this.provider,
+    );
+
+    const bundle = this.wallet.sign({ nonce, actions: [action] });
+    const result = await this.provider.aggregator.add(bundle);
+
+    if ("failures" in result) {
+      throw new Error(JSON.stringify(result.failures));
+    }
+
+    return this.constructTransactionResponse(
+      action,
+      result.hash,
+      this.wallet.address,
+      nonce,
+    );
+  }
+
+  async getAddress(): Promise<string> {
+    await this.initPromise;
+    if (this._address) {
+      return this._address;
+    }
+
+    this._address = this.wallet.address;
+    return this._address;
+  }
+
+  // Construct a response that follows the ethers TransactionResponse type
+  async constructTransactionResponse(
+    action: ActionData,
+    hash: string,
+    from: string,
+    nonce?: BigNumber,
+  ): Promise<ethers.providers.TransactionResponse> {
+    await this.initPromise;
+    const chainId = await this.getChainId();
+    if (!nonce) {
+      nonce = await BlsWalletWrapper.Nonce(
+        this.wallet.PublicKey(),
+        this.verificationGatewayAddress,
+        this.provider,
+      );
+    }
+
+    // TODO: bls-wallet #412 Update values returned in bundle receipt to more closely match ethers transaction response
+    return {
+      hash,
+      to: action.contractAddress,
+      from,
+      nonce: BigNumber.from(nonce).toNumber(),
+      gasLimit: BigNumber.from("0x0"),
+      data: action.encodedFunction.toString(),
+      value: BigNumber.from(action.ethValue),
+      chainId,
+      type: 2,
+      confirmations: 1,
+      wait: (confirmations?: number) => {
+        return this.provider.waitForTransaction(hash, confirmations);
+      },
+    };
+  }
+
+  async unlock(password: string): Promise<boolean> {
+    const provider = this.provider;
+
+    const address = await this.getAddress();
+
+    return provider.send("personal_unlockAccount", [
+      address.toLowerCase(),
+      password,
+      null,
+    ]);
+  }
+
+  override async signTransaction(
+    transaction: Deferrable<ethers.providers.TransactionRequest>,
+  ): Promise<string> {
+    await this.initPromise;
+
+    if (!transaction.to) {
+      throw new TypeError("Transaction.to should be defined.");
+    }
+
+    const action: ActionData = {
+      ethValue: transaction.value?.toString() ?? "0",
+      contractAddress: transaction.to.toString(),
+      encodedFunction: transaction.data?.toString() ?? "0x",
+    };
+
+    const nonce = await BlsWalletWrapper.Nonce(
+      this.wallet.PublicKey(),
+      this.verificationGatewayAddress,
+      this,
+    );
+
+    const bundle = this.wallet.sign({ nonce, actions: [action] });
+    return JSON.stringify(bundleToDto(bundle));
+  }
+
+  /** Sign a message */
+  override async signMessage(message: Bytes | string): Promise<string> {
```

Comment:
@jacque006 had a jab at updating the `signMessage` logic but unsure how correct it is. How does this look to you?

Reply:
Ok great, will leave for now and make a note to come back to this once EIP-1271 is supported